### PR TITLE
Add config for Tabular DP Model; deprecate differential-privacy

### PR DESCRIPTION
**Pull request by lipikaramaswamy**
- New config `tabular-differential-privacy.yml` for Tabular DP Model
- Deprecation notice for `differential-privacy.yml` for Tabular LSTM with differential privacy turned on.

### CODE REVIEW

File: config_templates/gretel/synthetics/differential-privacy.yml [hunk 1]
```diff
@@ -1,4 +1,8 @@
-# This is a default configuration for enabling differential privacy during
+# deprecated: This configuration will be deprecated soon. Please use the
+# "synthetics/tabular-lstm" configuration with dp: True to use experimental
```
[the remark this answers is not in the document]
Comment (kboyd) [reply config_templates/gretel/synthetics/differential-privacy.yml:2]: We resolved the batch size issue when you take the `synthetics/tabular-lstm` config and only change `dp: True`, right? So doing this won't throw an error like it would a few months ago?

File: config_templates/gretel/synthetics/tabular-differential-privacy.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Default configuration for Gretel Tabular DP to generate synthetic data with
+# differential privacy guarantees
```
Comment (kboyd) [inline config_templates/gretel/synthetics/tabular-differential-privacy.yml:2]:
```suggestion
# differential privacy guarantees.
```

File: config_templates/gretel/synthetics/tabular-differential-privacy.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Default configuration for Gretel Tabular DP to generate synthetic data with
```
Comment (kboyd) [inline config_templates/gretel/synthetics/tabular-differential-privacy.yml:1]:
```suggestion
# Default configuration for the Gretel Tabular DP model to generate synthetic data with
```

File: config_templates/gretel/synthetics/tabular-differential-privacy.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+# differential privacy guarantees
+
+schema_version: "1.0"
+name: "tabular-differential-privacy"
+models:
+  - tabular_dp:
+      data_source: __tmp__
+      params:
+        epsilon: 1
+        delta: 0.0001
```
Comment (kboyd) [inline config_templates/gretel/synthetics/tabular-differential-privacy.yml:11]: These are fairly aggressive params as defaults. Epsilon of 1 and expecting 10000 rows of input data. I'd guess most uses will be first-timers and with not that much data.

Does the quality drop that much from using `delta=0.0001` versus `delta=0.001` to make it worth being a bit more lax here? Or do you think it just looks better to have aggressive default params?

File: config_templates/gretel/synthetics/tabular-differential-privacy.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+        infer_domain: True
```
Comment (kboyd) [inline config_templates/gretel/synthetics/tabular-differential-privacy.yml:12]: I'd expect to see a newline at the end of the file (I think other yml files here do that, though a little hard to tell). Unless a VSCode or other autoformatter is removing a trailing newline automatically.

File: config_templates/gretel/synthetics/differential-privacy.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (lipikaramaswamy) [reply config_templates/gretel/synthetics/differential-privacy.yml:2]: Yes! This commit fixes it https://github.com/Gretellabs/monogretel/pull/1845

File: config_templates/gretel/synthetics/tabular-differential-privacy.yml (hunk 4 above)
Comment (lipikaramaswamy) [reply config_templates/gretel/synthetics/tabular-differential-privacy.yml:11]: The reason for the `eps=1, delta = 0.0001` choice is good results with benchmark, where I varied epsilon and kept `delta=0.0001` fixed.

- I noticed the improvement from increasing epsilon to 2 is marginal for most datasets (in fact, some datasets had better performance with eps=0.1). I wanted to highlight the particularly impressive performance because `eps=1` is expected to be an "academic" value, rather than a practical one.
- Of course, decreasing delta will help improve accuracy of the DP methods contained in this model. However, we'll always want to encourage setting this to `1/n`. I would advocate for implementing `delta: auto` very soon and updating the recommended value in this config.

File: config_templates/gretel/synthetics/differential-privacy.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (lipikaramaswamy) [reply config_templates/gretel/synthetics/differential-privacy.yml]: Just checked again - I was able to successfully kick off training by changing `dp: True` in the `tabular-lstm` config. Separate matter that there were too many invalid records.

```
15:32:20  Analyzing input data and checking for auto-params...

15:32:21  Found 7 auto-params that were set based on input data. — epochs 50, batch_size 4, outlier_filter null, similarity_filter null, vocab_size 0, rnn_units 256, learning_rate 0.001

15:32:21  Using updated model configuration: 

{

    "schema_version": "1.0",

    "name": "tabular-lstm",

    "models": [

        {

            "synthetics": {

                "privacy_filters": {

                    "outliers": null,

                    "similarity": null,

                    "max_iterations": 10

                },

                "data_source": [

                    "gretel_1c32aa58f4f4486f8055f9b9d71b49c8_train.csv"

                ],

                "ref_data": {},

                "params": {

                    "field_delimiter": null,

                    "epochs": 50,

                    "batch_size": 4,

                    "vocab_size": 0,

                    "reset_states": null,

                    "learning_rate": 0.001,

                    "rnn_units": 256,

                    "dropout_rate": null,

                    "overwrite": true,

                    "early_stopping": null,

                    "gen_temp": null,

                    "predict_batch_size": null,

                    "validation_split": false,

                    "dp": true,

                    "dp_noise_multiplier": null,

                    "dp_l2_norm_clip": null,

                    "dp_microbatches": null,

                    "early_stopping_min_delta": null,

                    "max_training_time_seconds": 28200,

                    "field_cluster_size": 20,

                    "average_record_length_threshold": 250.0,

                    "data_upsample_limit": 0

                },

                "validators": {

                    "in_set_count": 10,

                    "pattern_count": 10,

                    "datetime_formats": "infer",

                    "use_numeric_iqr": false,

                    "allow_empty_values": true,

                    "open_close_chars": null

                },

                "generate": {

                    "num_records": 5000,

                    "max_invalid": null

                },

                "task": null,

                "debug": {

                    "invalid_record_cutoff": 40

                },

                "data_checks": {

                    "strategy": "log"

                },

                "evaluate": {

                    "correlation_columns": 75,

                    "sqs_report_columns": 250,

                    "sqs_report_rows": 5000,

                    "mandatory_columns": null,

                    "target": null,

                    "holdout": null,

                    "metric": null,

                    "models": null,

                    "skip": false,

                    "task": null

                },

                "fallback": {

                    "allow": false,

                    "max_fallback_training_time_seconds": 300,

                    "max_fallback_generation_time_seconds": 300

                }

            }

        }

    ],

    "notifications": null,

    "label_predictors": null

}

15:32:38  Starting synthetic model training

15:32:38  Loading training data

15:32:38  Running pre-flight data checks on input data

15:32:41  We found 1 warning. If your model fails, try modifying the fields or adjusting your configuration settings.


	1 field with missing data: 'grapheme' has more than 50% missing data. 

15:32:41  Training data loaded. — record_count 200840, field_count 6, upsample_count 0

15:33:08  Creating semantic validators and preparing training data

15:33:27  Beginning ML model training

15:33:35  Running training on 1 batches. — batch_sizes [6]

15:33:37  Tokenizing input data

15:33:37  Shuffling input data

15:33:45  Initializing synthetic model

15:36:09  Training epoch complete. — epoch 0, accuracy 0.3339, loss 2.2114, val_accuracy 0, val_loss 0, batch 0, epsilon 362.5118, delta 0

15:38:27  Training epoch complete. — epoch 1, accuracy 0.5295, loss 1.604, val_accuracy 0, val_loss 0, batch 0, epsilon 514.5521, delta 0

15:40:49  Training epoch complete. — epoch 2, accuracy 0.5869, loss 1.3788, val_accuracy 0, val_loss 0, batch 0, epsilon 666.5924, delta 0

15:43:09  Training epoch complete. — epoch 3, accuracy 0.5965, loss 1.3238, val_accuracy 0, val_loss 0, batch 0, epsilon 818.6328, delta 0

15:45:35  Training epoch complete. — epoch 4, accuracy 0.5971, loss 1.3104, val_accuracy 0, val_loss 0, batch 0, epsilon 970.6731, delta 0

15:48:52  Training epoch complete. — epoch 5, accuracy 0.5993, loss 1.2992, val_accuracy 0, val_loss 0, batch 0, epsilon 1122.7134, delta 0

15:51:14  Training epoch complete. — epoch 6, accuracy 0.599, loss 1.2962, val_accuracy 0, val_loss 0, batch 0, epsilon 1274.7538, delta 0

15:53:43  Training epoch complete. — epoch 7, accuracy 0.5991, loss 1.2959, val_accuracy 0, val_loss 0, batch 0, epsilon 1426.7941, delta 0

15:56:05  Training epoch complete. — epoch 8, accuracy 0.5999, loss 1.2894, val_accuracy 0, val_loss 0, batch 0, epsilon 1578.8344, delta 0

15:59:27  Training epoch complete. — epoch 9, accuracy 0.5991, loss 1.2932, val_accuracy 0, val_loss 0, batch 0, epsilon 1730.8748, delta 0

16:01:52  Training epoch complete. — epoch 10, accuracy 0.5998, loss 1.291, val_accuracy 0, val_loss 0, batch 0, epsilon 1882.9151, delta 0

16:04:14  Training epoch complete. — epoch 11, accuracy 0.5989, loss 1.2892, val_accuracy 0, val_loss 0, batch 0, epsilon 2034.9554, delta 0

16:06:35  Training epoch complete. — epoch 12, accuracy 0.5969, loss 1.2943, val_accuracy 0, val_loss 0, batch 0, epsilon 2186.9957, delta 0

16:09:02  Training epoch complete. — epoch 13, accuracy 0.5985, loss 1.2916, val_accuracy 0, val_loss 0, batch 0, epsilon 2339.0361, delta 0

16:09:02  Training LSTM model completed successfully.

16:09:02  Generating records — num_records 5000, max_invalid 10000

16:09:07  Generation in progress — current_valid_count 0, current_invalid_count 0, new_valid_count 0, new_invalid_count 0, completion_percent 0

16:09:12  Generation in progress — current_valid_count 0, current_invalid_count 0, new_valid_count 0, new_invalid_count 0, completion_percent 0

16:09:17  Generation in progress — current_valid_count 0, current_invalid_count 0, new_valid_count 0, new_invalid_count 0, completion_percent 0

16:09:22  Generation in progress — current_valid_count 0, current_invalid_count 0, new_valid_count 0, new_invalid_count 0, completion_percent 0

16:09:27  Generation in progress — current_valid_count 0, current_invalid_count 0, new_valid_count 0, new_invalid_count 0, completion_percent 0

16:09:32  Generation in progress — current_valid_count 0, current_invalid_count 0, new_valid_count 0, new_invalid_count 0, completion_percent 0

16:09:37  Generation in progress — current_valid_count 0, current_invalid_count 0, new_valid_count 0, new_invalid_count 0, completion_percent 0

16:09:42  Generation in progress — current_valid_count 17, current_invalid_count 1738, new_valid_count 17, new_invalid_count 1738, completion_percent 0.34

16:09:47  Generation in progress — current_valid_count 45, current_invalid_count 3880, new_valid_count 28, new_invalid_count 2142, completion_percent 0.9

16:09:52  Generation in progress — current_valid_count 69, current_invalid_count 6016, new_valid_count 24, new_invalid_count 2136, completion_percent 1.38

16:09:57  Generation in progress — current_valid_count 96, current_invalid_count 8119, new_valid_count 27, new_invalid_count 2103, completion_percent 1.92

16:10:02  Generation in progress — current_valid_count 115, current_invalid_count 10001, new_valid_count 19, new_invalid_count 1882, completion_percent 2.3

16:10:02  Generation failure, gathering invalid record logs — validation_logs [object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]

  


======== ERROR

Could not generate all requested records. Valid records: 115, Invalid records: 10001. 

```

File: config_templates/gretel/synthetics/tabular-differential-privacy.yml (hunk 4 above)
Comment (kboyd) [reply config_templates/gretel/synthetics/tabular-differential-privacy.yml:11]: Excellent, using `eps=1` as the default is great. And sounds like you can add the `delta: auto` support soon.

**Review by kboyd — APPROVED** [review]
